Update the postprocessing service (#358)
owncloud/docs-ocis: modules/ROOT/pages/deployment/services/s-list/postprocessing.adoc

== General Prerequisites

> To use the postprocessing service, an event system needs to be configured for all services. By default, Infinite Scale ships with a preconfigured `nats` service.

Reviewer: ```suggestion
To use the post-processing service, an event system needs to be configured for all services. By default, Infinite Scale ships with a preconfigured `nats` service.
```

> == Postprocessing Functionality

Reviewer: ```suggestion
== Post-Processing Functionality
```

> == Postprocessing Steps

Reviewer: ```suggestion
== Post-Processing Steps
```

> The postporcessing service is individually configurable. This is achieved by allowing a list of postprocessing steps that are processed in order of their appearance in the `POSTPROCESSING_STEPS` envvar. This envvar expects a comma separated list of steps that will be executed. Currently known steps to the system are `virusscan` and `delay`. Custom steps can be added but need an existing target for processing.

Reviewer: ```suggestion
The post-processing service is individually configurable. This is achieved by allowing a list of post-processing steps to be performed in order of their appearance in the `POSTPROCESSING_STEPS` envvar. This envvar expects a comma-separated list of steps that will be executed. Currently known steps to the system are `virusscan` and `delay`. Custom steps can be added but need an existing target for processing.
```

=== Virus Scanning

> To enable virus scanning as a postprocessing step after uploading a file, the environment variable `POSTPROCESSING_STEPS` needs to contain the word `virusscan` at one location in the list of steps. As a result, each uploaded file gets virus scanned as part of the postprocessing steps. Note that the `antivirus service` is required to be enabled and configured for this to work.

Reviewer: ```suggestion
To enable virus scanning as a post-processing step after uploading a file, the environment variable `POSTPROCESSING_STEPS` needs to contain the word `virusscan` at one location in the list of steps. As a result, each uploaded file gets virus-scanned as part of the post-processing steps. Note that the `antivirus service` must be enabled and configured for this to work.
```

=== Delay

> Though this is for development purposes only and NOT RECOMMENDED on production systems, setting the environment variable `POSTPROCESSING_DELAY` to a duration not equal to zero will add a delay step with the configured amount of time. Infinite Scale will continue postprocessing the file after the configured delay. Use the enviroment variable `POSTPROCESSING_STEPS` and the keyword `delay` if you have multiple postprocessing steps and want to define their order. If `POSTPROCESSING_DELAY` is set but the keyword `delay` is not contained in `POSTPROCESSING_STEPS`, it will be processed as last postprocessing step without being listed there. In this case, a log entry will be written on service startup to notify the admin about that situation. That log entry can be avoided by adding the keyword `delay` to `POSTPROCESSING_STEPS`.

Reviewer: ```suggestion
Though this is for development purposes only and NOT RECOMMENDED on production systems, setting the environment variable `POSTPROCESSING_DELAY` to a duration not equal to zero will add a delay step with the configured amount of time. Infinite Scale will continue post-processing the file after the configured delay. Use the enviroment variable `POSTPROCESSING_STEPS` and the keyword `delay` if you have multiple post-processing steps and want to define their order. If `POSTPROCESSING_DELAY` is set but the keyword `delay` is not contained in `POSTPROCESSING_STEPS`, it will be executed as the last post-processing step without being listed as the last one. In this case, a log entry will be written on service startup to notify the admin about the situation. That log entry can be avoided by adding the keyword `delay` to `POSTPROCESSING_STEPS`.
```

> == Custom Postprocessing Steps

Reviewer: ```suggestion
== Custom Post-Processing Steps
```

> By using the envvar `POSTPROCESSING_STEPS`, custom postprocessing steps can be added. Any word can be used as step name but be careful not to conflict with exising keywords like `virusscan` and `delay`. In addition, if a keyword is misspelled or the corresponding service does either not exist or does not follow the necessary event communication, the postprocessing service will wait forever getting the required response to proceed and does not continue any other processing.

Reviewer: ```suggestion
By using the envvar `POSTPROCESSING_STEPS`, custom post-processing steps can be added. Any word can be used as step name but be careful not to conflict with existing keywords like `virusscan` and `delay`. In addition, if a keyword is misspelled or the corresponding service does either not exist or does not follow the necessary event communication, the post-processing service will wait forever to get the required response to proceed and therefore does not continue with any other processing.
```

=== Prerequisites

> For using custom postprocessing steps, you need a custom service listening to the configured event system, see xref:general-prerequisites[General Prerequisites]

Reviewer: ```suggestion
To use custom post-processing steps, you need a custom service listening to the configured event system. For more information, see xref:general-prerequisites[General Prerequisites].
```

=== Workflow

> When setting a custom postprocessing step (eg. "customstep") the postprocessing service will eventually sent an event during postprocessing. The event will be of type StartPostprocessingStep with its field StepToStart set to "customstep". When the custom service receives this event, it can savely execute its actions. The postprocessing service will wait until it has finished its work. The event contains further information (like filename, executing user, size, ...) and required tokens respectively urls to download the file in case byte inspection is necessary.

Reviewer: ```suggestion
When setting a custom post-processing step (eg. "customstep") the post-processing service will eventually send an event during post-processing. The event will be of the type StartPostprocessingStep with its field StepToStart set to "customstep". When the custom service receives this event, it can safely execute its actions. The post-processing service will wait until it has finished its work. The event contains further information (like filename, executing user, size, ...) and required tokens respectively URLs to download the file in case byte inspection is necessary.
```

> Once the custom service has finished its work, it should sent an event of type `PostprocessingFinished` via the configured events system. This event needs to contain a `FinishedStep` field set to "customstep". It also must contain the outcome of the step, which can be one of `delete` (abort postprocessing, delete the file), `abort` (abort postprocessing, keep the file) and `continue` (continue postprocessing, this is the success case).

Reviewer: ```suggestion
Once the custom service has finished its work, it should send an event of the type `PostprocessingFinished` via the configured event system. This event needs to contain a `FinishedStep` field set to "customstep". It also must contain the outcome of the step, which can be one of `delete` (abort post-processing, delete the file), `abort` (abort post-processing, keep the file) and `continue` (continue post-processing, this is the success case).
```
